qubesadmin/tools/qvm_ls.py

```diff
@@ -594,60 +626,113 @@ def get_parser():
-
-    for pwrstate in DOMAIN_POWER_STATES:
-        parser.add_argument('--{}'.format(pwrstate), action='store_true',
-            help='show {} VMs'.format(pwrstate))
-
-    parser.add_argument('--raw-data', action='store_true',
-        help='Display specify data of specified VMs. Intended for '
-             'bash-parsing.')
-
-    parser.add_argument('--tree', '-t',
+    parser_format.add_argument('--tree', '-t',
         action='store_const', const='tree',
         help='sort domain list as network tree')
 
-    parser.add_argument('--spinner',
-        action='store_true', dest='spinner',
-        help='reenable spinner')
-
-    parser.add_argument('--no-spinner',
-        action='store_false', dest='spinner',
-        help='disable spinner')
+    parser_format.add_argument('--raw-data', action='store_true',
+        help='Display specify data of specified VMs. Intended for '
+             'bash-parsing.')
 
     # shortcuts, compatibility with Qubes 3.2
-    parser.add_argument('--raw-list', action='store_true',
+    parser_format.add_argument('--raw-list', action='store_true',
         help='Same as --raw-data --fields=name')
 
-    parser.add_argument('--disk', '-d',
+    parser_format_exclusive.add_argument('--disk', '-d',
         action='store_const', dest='format', const='disk',
         help='Same as --format=disk')
 
-    parser.add_argument('--network', '-n',
+    parser_format_exclusive.add_argument('--network', '-n',
         action='store_const', dest='format', const='network',
         help='Same as --format=network')
 
-    parser.add_argument('--kernel', '-k',
+    parser_format_exclusive.add_argument('--kernel', '-k',
         action='store_const', dest='format', const='kernel',
         help='Same as --format=kernel')
 
+    parser_format.add_argument('--help-formats', action=_HelpFormatsAction)
+    parser_format.add_argument('--help-columns', action=_HelpColumnsAction)
+
+    parser_filter = parser.add_argument_group(title='filtering options')
+
+    parser_filter.add_argument('--class', nargs='+', metavar='CLASS',
+        dest='klass', action='store',
+        help='show only qubes of specific class(es)')
+
+    parser_filter.add_argument('--label', nargs='+', metavar='LABEL',
+        action='store',
+        help='show only qubes with specific label(s)')
+
+    parser_filter.add_argument('--tags', nargs='+', metavar='TAG',
+        help='show only qubes having specific tag(s)')
+
+    parser_filter.add_argument('--exclude-tags', nargs='+', metavar='TAG',
+        help='exclude qubes having specific tag(s)')
+
+    for pwstate in DOMAIN_POWER_STATES:
+        parser_filter.add_argument('--{}'.format(pwstate), action='store_true',
+        help='show {} VMs'.format(pwstate))
+
+    parser_filter.add_argument('--template-source', nargs='+',
+        metavar='TEMPLATE', action='store',
+        help='filter results to the qubes based on the TEMPLATE. '
+             '"" means None')
```

Comment:
> Omitted VALUE means None (not set). "" or '' means blank

What about:

_Omitted VALUE means None (unset). Empty value means "" or '' (blank)_

Reply:
This one is tricky. A template name could not be blank. This is a case that we could have only None.